Allow for separate Network resource group

frobware opened the pull request:
Allow for separate Network resource group

The goal is to allow networking components like the subnets, VNET to
be in a separate resource group but the resources for the cluster like
LBs, network interfaces, VMs, etc stay in the resource group of the
cluster.

The following services remain in the existing `ResourceGroup`:
- availabilityzones
- certificates
- config
- disks
- groups
- internalloadbalancers
- networkinterfaces
- publicips
- publicloadbalancers
- routetables
- securitygroups
- virtualmachineextensions
- virtualmachines

The following services are now in the new `NetworkResourceGroup`:
- subnets
- virtualnetworks

https://jira.coreos.com/browse/CLOUD-594

frobware commented:
/hold

WIP

frobware commented:
/test e2e-azure

frobware commented:
CC @abhinavdahiya

frobware commented:
/retest

abhinavdahiya commented on pkg/cloud/azure/services/internalloadbalancers/internalloadbalancers.go:
```diff
@@ -62,7 +62,7 @@ func (s *Service) CreateOrUpdate(ctx context.Context, spec azure.Spec) error {
 	frontEndIPConfigName := "controlplane-internal-lbFrontEnd"
 	backEndAddressPoolName := "controlplane-internal-backEndPool"
-	idPrefix := fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/loadBalancers", s.Scope.SubscriptionID, s.Scope.ClusterConfig.ResourceGroup)
+	idPrefix := fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/loadBalancers", s.Scope.SubscriptionID, s.Scope.ClusterConfig.NetworkResourceGroup)
```
The load balancers are part of the cluster so they should stay in the cluster's resource group.. right?

frobware replied:
Yes; updated in 9f037f5...76ceed6.

abhinavdahiya commented:
The goal is to allow networking components like the subnets, VNET be in a separate resource group... But the resources for the cluster like lbs network interfaces VMs etc stay in the resource group of the cluster...

frobware commented:
/test e2e-azure

frobware commented:
> The goal is to allow networking components like the subnets, VNET be in a separate resource group... But the resources for the cluster like lbs network interfaces VMs etc stay in the resource group of the cluster...

I updated the description with a list of services that should remain in the existing `ResourceGroup` and those that should be in the new `NetworkResourceGroup`. I also deleted my changes to the `servives/*.go` files. Once we have agreed on the split I will resurrect the changes. TIA.

ingvagabund commented:
@abhinavdahiya can you make a list of all resource that are to be considered networking components as you suggest? Is it just subnets and VNET? Or, are there others?

abhinavdahiya commented:
> @abhinavdahiya can you make a list of all resource that are to be considered networking components as you suggest? Is it just subnets and VNET? Or, are there others? The PR description provides a list of all resources the azure actuator is aware of.

VNET and subnets for now. The description is very helpful thanks

ingvagabund commented:
FYI: https://github.com/Azure/azure-cli/issues/2387

ingvagabund commented:
I am not able to create an interface in one group with subnet in another. Still getting:
```
Resource /subscriptions/6648228a-736c-4a4e-9a14-92a7e5190f4a/resourceGroups/os4-common-network/providers/Microsoft.Network/virtualNetworks/os4-common-network-vnet/subnets/os4-common-network-subnet referenced by resource /subscriptions/6648228a-736c-4a4e-9a14-92a7e5190f4a/resourceGroups/os4-common/providers/Microsoft.Network/networkInterfaces/master-machine-nic was not found. Please make sure that the referenced resource exists, and that both resources are in the same region.
```

The subnet ID string is genererated only when it exists, so both network resource group and the virtual network resource names are correctly specified. Just wondering, what else is missing that makes the Azure think the subnet resource in another resource group does not exist.

ingvagabund commented:
I was able to create VM with cluster resources in one RG and vnet and its subnet in another RG. Also validating the case where RG is the same for both cluster and network.

/approve
/lgtm

frobware commented:
> I was able to create VM with cluster resources in one RG and vnet and its subnet in another RG. Also validating the case where RG is the same for both cluster and network.

Thanks for validating this.

frobware commented:
/hold cancel
